src/app/invest/invest.component.ts

```diff
@@ -10,68 +9,55 @@ import { map, startWith, tap } from 'rxjs/operators'
-    maxInvestment: 100000,
-    imgSrc: "https://images.pexels.com/photos/2850347/pexels-photo-2850347.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260"
+    minInvestment: 100,
+    maxInvestment: 1000,
+    imgSrc: "https://images.pexels.com/photos/2850347/pexels-photo-2850347.jpeg?auto=compress&cs=tinysrgb&dpr=2&h=750&w=1260",
   })
+  project$ = this.projectSub.asObservable()
+  investStateSub = new BehaviorSubject<InvestState>(InvestState.Editing)
+  investState$ = this.investStateSub.asObservable()
 
-  project$                        = this.project.asObservable()
-  componentState                  = new BehaviorSubject<ComponentState>(ComponentState.Editing)
-  investmentAmountForm            = this.fb.group({ investmentAmount: ['', Validators.required] })
-  investmentAmountValueChanges$   = this.investmentAmountForm.controls['investmentAmount'].valueChanges
-  investAmountStateType           = InvestmentAmountState
+  investmentForm: FormGroup
 
-  investmentAmountState$ =
-    this.investmentAmountValueChanges$
-      .pipe(map((amount) => this.handleAmountValueChange(parseInt(amount)) ), 
-            startWith(this.handleAmountValueChange(0)))
-
-  investmentAmount$ = 
-    this.investmentAmountValueChanges$.pipe(map((amount) => parseInt(amount) ))
-
-  isInReview$ = 
-    this.componentState.asObservable().pipe(
-      map((x) => x == ComponentState.InReview ))
-
-  isNextButtonDisabled$ = this.investmentAmountState$.pipe(
-    map((state) => state !== InvestmentAmountState.Valid ))
-
-  constructor(private fb: FormBuilder) { }
-
-  handleAmountValueChange(amount: number): InvestmentAmountState {
-
-    let model = this.project.value
-    let investmentAmountFieldIsEmpty = this.investmentAmountForm.controls['investmentAmount'].value !== ''
-
-    if (!investmentAmountFieldIsEmpty) { return InvestmentAmountState.Empty }
-    if (amount < model.minInvestment) { return InvestmentAmountState.InvestmentAmountTooLow }
-    if (amount > model.maxInvestment) { return InvestmentAmountState.InvestmentAmountTooHigh }
-    if (amount > model.walletBalance) {  return InvestmentAmountState.NotEnoughFunds }
-
-    return InvestmentAmountState.Valid
+  constructor(private fb: FormBuilder) {
+    this.investmentForm = this.fb.group({
+      amount: [0, [this.amountValidator()]],
+    })
   }
 
-  nextButtonClicked() {
-    this.componentState.next(ComponentState.InReview)
-    this.setAmountInputEnabled(false)
+  private amountValidator(): ValidatorFn {
+    return (control: AbstractControl) => {
+      const project = this.projectSub.value
+      const amount = control.value
+
+      if (!amount) {
+        return {amountEmpty: true}
+      } else if (amount < project.minInvestment) {
+        return {amountTooLow: true}
+      } else if (amount > project.maxInvestment) {
+        return {amountTooHigh: true}
+      } else if (amount > project.walletBalance) {
+        return {walletBalanceTooLow: true}
+      }
+
+      return null
+    }
```

Comment:
Take a look here how validators should be used. (a [random](https://blog.thoughtram.io/angular/2016/03/14/custom-validators-in-angular-2.html) article I found about it.

It's a simple function that accepts control as a parameter and returns `{[key: string]: true}` if there is an error on this control, otherwise `null`. Just inject it in the form definition and you're all done. No need for exhausting signaling with observables about the values. If something is invalid, your form will be invalid and have these errors attached to your value, and then you deal with errors inside templates. As a result, there is no need for the pipe you used.

Reply:
btw. there are more tricks with async validators where you can, for example, try to validate the field after some debounce period (e.g. 300ms) so the user doesn't get an error as he/she/it types. It's the third parameter inside a form builder object value and returns a promise or observable. we'll add it later if we need this, just to let you know.